apps/onnx/onnx_converter.cc

```diff
@@ -2682,6 +2831,116 @@ Node convert_shrink_node(
+                throw std::invalid_argument(
+                    "Attribute size should be > 0 but its " + std::to_string(size));
+            }
+            found_size = true;
+        }
+    }
+
+    if (!found_size) {
+        throw std::invalid_argument(
+            "Attribute size is required for lrn node " + node.name());
+    }
+
+    const Tensor &input = inputs[0];
+    Node result;
+    result.inputs = inputs;
+    result.outputs.resize(1);
+    result.outputs[0].shape = input.shape;
+    result.outputs[0].type = input.type;
+    result.outputs[0].rep = func_for_node_output(node, 0);
+
+    Halide::Var n("n"), c("c");
+    Halide::Func sum_squares(node.name() + "_sum_squares");
+
+    Halide::RDom r(-div_down(size - 1, 2), size);
+
+    if (input.shape.size() < 2) {
+        throw std::invalid_argument(
+            "Input rank must be at least 2 but its " +
+            std::to_string(input.shape.size()));
+    }
+
+    Halide::Expr c_size = input.shape[1];
+
+    sum_squares(n, c, Halide::_) = Halide::sum(Halide::select(
+        c + r < 0 || c + r >= c_size,
+        0,
+        input.rep(n, Halide::clamp((c + r), 0, c_size - 1), Halide::_) *
+            input.rep(n, Halide::clamp((c + r), 0, c_size - 1), Halide::_)));
+
+    result.outputs[0].rep(n, c, Halide::_) = input.rep(n, c, Halide::_) /
+                                             (Halide::pow(bias + (alpha / size) * sum_squares(n, c, Halide::_), beta));
+
+    return result;
+}
+
+Node convert_isinf_node(
+    const onnx::NodeProto &node,
+    const std::vector<Tensor> &inputs) {
+    if (inputs.size() != 1) {
+        throw std::invalid_argument(
+            "Expected exactly one input for isinf node " + node.name());
+    }
+
+    bool detect_negative = true;
+    bool detect_positive = true;
+    for (const auto &attr : node.attribute()) {
+        if (attr.name() == "detect_negative") {
+            detect_negative = attr.i();
+        } else if (attr.name() == "detect_positive") {
+            detect_positive = attr.i();
+        }
+    }
+
+    const Tensor &input = inputs[0];
+    Node result;
+    result.inputs = inputs;
+    result.outputs.resize(1);
+    result.outputs[0].shape = input.shape;
+    result.outputs[0].type = onnx::TensorProto_DataType_BOOL;
+    result.outputs[0].rep = func_for_node_output(node, 0);
+    if (input.type == onnx::TensorProto_DataType_FLOAT ||
+        input.type == onnx::TensorProto_DataType_DOUBLE) {
+        Halide::Expr inf_value = input.type == onnx::TensorProto_DataType_FLOAT
+                                     ? Halide::Expr(std::numeric_limits<float>::infinity())
+                                     : Halide::Expr(std::numeric_limits<double>::infinity());
+
+        Halide::Expr pos_inf = detect_positive && input.rep(Halide::_) == inf_value;
+        Halide::Expr neg_inf =
+            detect_negative && input.rep(Halide::_) == -inf_value;
+        // TODO(ataei): Fix Halide's mapping of const nanf into inf_f32.
```

Comment:
Does this need an issue to be opened to track?

Reply:
I have filed #4100 to request the ability to check for infinite values to be added to Halide proper. It's a much better place for it. I['m not sure about the actual TODO though: @asaadaldien should comment here.

Reply:
This TODO was investigate why evaluating (Halide::Expr == INFINITY) depends on the type ?! e.g in the following example f1, f2 should have the same values but f2 will have 1s wherever g have nanf.
```
f1(halide_var) = Halide::cast<float>(g(halide_var) == INFINITY); 
f2(halide_var) = Halide::cast<int>(g(halide_var) == INFINITY); 
```

A slightly different repro below : 
```
#include <math.h>
#include <Halide.h>

int main(int argc, char** argv) {
  float data[4] = {NAN, INFINITY, -INFINITY, 1.0f};
  Halide::Buffer<float> buff(&data[0], 4);
  Halide::Func f, g;
  Halide::Var x;
  f(x) = Halide::select(buff(x) == INFINITY, 1.0f, buff(x) == -INFINITY, -1.0f, 0.0f);
  g(x) = Halide::select(buff(x) == INFINITY, 1, buff(x) == -INFINITY, -1, 0);
  Halide::Buffer<float> f_out = f.realize(4);
  printf("f = [%.1f, %.1f, %.1f, %.1f], expected = [0.0, 1.0, -1.0, 0.0] \n", f_out(0), f_out(1), f_out(2), f_out(3));
  Halide::Buffer<int> g_out = g.realize(4);
  printf("g = [%d, %d, %d, %d], expected = [0, 1, -1, 0] \n", g_out(0), g_out(1), g_out(2), g_out(3));
  return 0;
}
```